### PR TITLE
[MetadataViewer] Introduce Metadata Viewer

### PR DESCRIPTION
**Pull request by jihong-yu**
this commit adds Metadata Viewer

ONE-vscode-DCO-1.0-Signed-off-by: jihong-yu <jimdac@naver.com>

### CODE REVIEW

**Comment by dayo09** [comment]
@jihong-yu Could you format this files using 'eslint' and 'clang-format' ?

File: src/OneExplorer/OneExplorer.ts [hunk 1]
```diff
@@ -137,7 +137,7 @@ enum NodeType {
   product,
 }
 
-abstract class Node {
+export abstract class Node {
```
Comment (dayo09) [inline src/OneExplorer/OneExplorer.ts:140]: Q. Is this change required?

File: package.json [hunk 2]
```diff
@@ -97,6 +97,26 @@
           }
         ],
         "priority": "option"
+      },
+      {
+        "viewType": "one.viewer.metadata",
+        "displayName": "Metadata",
+        "selector": [
+          {
+            "filenamePattern": "*"
```
Comment (dayo09) [inline package.json:106]: Could you specify the target filename patterns?

File: package.json [hunk 3]
```diff
@@ -97,6 +97,26 @@
...
+          }
+        ],
+        "priority": "option"
+      },
+      {
+        "viewType": "one.viewer.relation",
+        "displayName": "Relation",
+        "selector": [
+          {
+            "filenamePattern": "*"
```
Comment (dayo09) [inline package.json:116]: Could you specify the target filename patterns, too?

File: package.json [hunk 4]
```diff
@@ -97,6 +97,26 @@
           }
         ],
         "priority": "option"
+      },
+      {
+        "viewType": "one.viewer.metadata",
+        "displayName": "Metadata",
```
Comment (dayo09) [inline package.json:103]:
```suggestion
        "displayName": "Metadata Viewer",
```

File: package.json [hunk 5]
```diff
@@ -97,6 +97,26 @@
...
+        "selector": [
+          {
+            "filenamePattern": "*"
+          }
+        ],
+        "priority": "option"
+      },
+      {
+        "viewType": "one.viewer.relation",
+        "displayName": "Relation",
```
Comment (dayo09) [inline package.json:113]:
```suggestion
        "displayName": "Relation Viewer",
```

File: package.json [hunk 6]
```diff
@@ -195,6 +215,21 @@
         "command": "one.viewer.jsonTracer",
         "title": "json tracer",
         "category": "ONE"
+      },
+      {
+        "command": "one.viewer.metadata.show",
```
Comment (dayo09) [inline package.json:220]: Please add those commands to "commandPalette" and set "false".

For example, L399 of `package.json`,
image


This will conceal those commands from command palette.
Currently, they are shown as below. 
image

**Comment by dayo09** [comment]
### Screenshot

image


Thanks, it works very well with the example file. :-D

File: package.json (hunk 5 above)
Comment (dayo09) [inline package.json:113]: Could you remove these `Relation Viewer`-related lines? (1)

File: package.json [hunk 7]
```diff
@@ -195,6 +215,21 @@
...
+        "title": "Show Metadata",
+        "category": "ONE"
+      },
+      {
+        "command": "one.viewer.relation.show",
```
Comment (dayo09) [inline package.json:225]: Could you remove these `Relation Viewer`-related lines? (2)

File: src/OneExplorer/OneExplorer.ts (hunk 1 above)
Comment (jihong-yu) [reply src/OneExplorer/OneExplorer.ts:140]: It is necessary to distinguish whether the viewer was executed in ONE or in Explorer. it is classified as an instance of a node.  However, it may be classified in other ways. Should we not export as much as possible?

File: package.json (hunk 2 above)
Comment (jihong-yu) [reply package.json:106]: There are five types of files that we need to set. But I don't know how to set up multiple files. It is not allowed to be treated in an array. Therefore, all file patterns were given as "priority": "option" and processed so that the viewer could only be opened in a certain file.
("when": "view == OneExplorerView && viewItem =~ /baseModel|product/")

File: package.json (hunk 4 above)
Comment (jihong-yu) [reply package.json:103]: All right, I'll change it to that name.

File: package.json (hunk 6 above)
Comment (jihong-yu) [reply package.json:220]: All right, I'll apply that part.

File: package.json (hunk 5 above)
Comment (jihong-yu) [reply package.json:113]: All right, I'll apply that part.

File: package.json (hunk 7 above)
Comment (jihong-yu) [reply package.json:225]: All right, I'll apply that part.

File: package.json [hunk 8]
```diff
@@ -341,6 +368,10 @@
         {
           "command": "one.viewer.jsonTracer",
           "when": "false"
+        },
+        {
+          "command": "one.viewer.metadata.show",
```
Comment (dayo09) [inline package.json:373]: Thanks :-D

File: src/OneExplorer/OneExplorer.ts (hunk 1 above)
Comment (dayo09) [reply src/OneExplorer/OneExplorer.ts]: > Should we not export as much as possible?

I just wondered why it is exported. It looks that you have found out the other way and excluded this change.

File: package.json [hunk 9]
```diff
@@ -195,6 +205,11 @@
         "command": "one.viewer.jsonTracer",
         "title": "json tracer",
         "category": "ONE"
+      },
+      {
+        "command": "one.viewer.metadata.show",
+        "title": "Show Metadata",
+        "category": "ONE"
       }
```
Comment (dayo09) [inline package.json:213]:
```
      {
        "command": "one.viewer.metadata.showInDefaultExplorer",
        "title": "ONE: Show Metadata",
        "category": "ONE"
      }
```

File: src/MetadataViewer/MetadataViewerProvider.ts [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+      vscode.window.registerCustomEditorProvider(MetadataViewerProvider.viewType, provider, {
+        webviewOptions: {
+          retainContextWhenHidden: true,
+        }
+      }),
+      vscode.commands.registerCommand(
+          'one.viewer.metadata.show',
+          async (uri) => {
+            // If the method is executed in the ONE Explorer, change the uri instance.
+            let fileUri = uri.uri ? uri.uri : uri;
```
Comment (dayo09) [inline src/MetadataViewer/MetadataViewerProvider.ts:98]: Could you separate 'one.viewer.metadata.show' command?

It does show up in both `context menu in default explorer` and `context menu in ONE Explorer`.

How about adding another command for `default vscode explorer`, like 'one.viewer.metadata.showInDefaultExplorer'?
Plus, if you do so, would you put the command title as 'ONE: Show Metadata', instead of 'Show Metadata' for the new command? Like, https://github.com/Samsung/ONE-vscode/pull/1315/files#r985182428

Comment (jihong-yu) [reply src/MetadataViewer/MetadataViewerProvider.ts:98]: ok!!, I agree this good suggestion. I`am going to apply this part right now.

File: package.json (hunk 9 above)
Comment (jihong-yu) [reply package.json:213]: thank you, I think it became clearer with ONE on "Show Metadata".

**Review by dayo09 — APPROVED** [review]
LGTM, Thanks!

File: package.json [hunk 11]
```diff
@@ -195,6 +205,16 @@
...
         "category": "ONE"
+      },
+      {
+        "command": "one.viewer.metadata.showFromDefaultExplorer",
+        "title": "ONE: Show Metadata",
+        "category": "ONE"
+      },
+      {
+        "command": "one.viewer.metadata.showFromOneExplorer",
+        "title": "ONE: Show Metadata",
```
Comment (dayo09) [inline package.json:216]: How about removing this `ONE:` if it's a context menu of ONE Explorer?

Comment (jihong-yu) [reply package.json:216]: ok, I got it! thank you!

**Review by dayo09 — DISMISSED** [review]
LGTM, Thanks!

**Review by dayo09 — DISMISSED** [review]

**Comment by dayo09** [comment]
[[Metadata Viewer] introduce metadata viewer](https://github.com/Samsung/ONE-vscode/pull/1315/commits/07d1384c42180ca0c1b8adcdb9ccd86ab5eefd0b)

Could you update the commit title as'[MetadataViewer] Introduce Metadata Viewer'

**Review by smj53 — APPROVED** [review]

**Comment by llFreetimell** [comment]
From second commit, there is no need to add signed-off-by and title :)

**Comment by dayo09** [comment]
@llFreetimell I think it's by adding a git-hook, just automatically generated... (Same thing happens to me) Could you please check the commit title when you merge this PR?

**Comment by llFreetimell** [comment]
First commit's title is used :)
Maybe `git commit -m "message"` would work...?

Anyway, no problem to merge :)

**Comment by dayo09** [comment]
I mean, I changed the PR title but not the commit title, so I asked you to watch out that the current title is being merged.. 
+ I use the `git commit -m` command but still it put the sign-off ☹️

**Comment by dayo09** [comment]
@jihong-yu Could you remove those sign-off-by lines of each commit, except for the first commit?

**Comment by dayo09** [comment]
## GIF
1001-metadata-viewer

**Review by dayo09 — DISMISSED** [review]
LGTM, Thanks!

**Comment by dayo09** [comment]
@smj53 PTAL :-D

**Review by smj53 — DISMISSED** [review]

File: media/MetadataViewer/style.css [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+ * limitations under the License.
+ */
+
+body {
+  background-color: #1e1e1e;
+  max-width: 800px;
+  margin-left: 20px;
+}
+
+/* white테마일 때 처리 */
```
Comment (llFreetimell) [inline media/MetadataViewer/style.css:23]: Please change this to English :)

File: src/MetadataViewer/example/MetadataExample.ts [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+NOTE this is example logic is that you can get sample MetaData
```
Comment (llFreetimell) [inline src/MetadataViewer/example/MetadataExample.ts:18]:
```suggestion
NOTE this is example logic that you can get sample MetaData
```

File: src/MetadataViewer/MetadataViewerProvider.ts [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+  private constructor(uri: vscode.Uri) {
+    this._uri = uri;
+    this._metadataViwer = [];
+  }
+
+  public get uri() {
+    return this._uri;
+  }
+
+  // CustomDocument implements
```
Comment (llFreetimell) [inline src/MetadataViewer/MetadataViewerProvider.ts:41]: Maybe mismatched comment..?

File: media/MetadataViewer/style.css (hunk 12 above)
Comment (jihong-yu) [reply media/MetadataViewer/style.css:23]: Oh, I'm sorry. I'm going to fix it now.

File: src/MetadataViewer/example/MetadataExample.ts (hunk 13 above)
Comment (jihong-yu) [reply src/MetadataViewer/example/MetadataExample.ts:18]: Oh, I'm sorry. I'm going to fix it now.

**Review by dayo09 — APPROVED** [review]

**Review by smj53 — APPROVED** [review]

**Review by llFreetimell — APPROVED** [review]
